core/src/avm2/globals/flash/utils/bytearray.rs

```diff
@@ -0,0 +1,946 @@
+            .unwrap()
+            .as_bytearray_mut(activation.context.gc_context)
+        {
+            bytearray.write_bytes(if length != 0 {
+                &combining_bytes[offset..length + offset]
+            } else {
+                &combining_bytes[offset..]
+            });
+        }
+    }
+
+    Ok(Value::Undefined)
+}
+
+// Reads the bytes from the current bytearray into another bytearray
+pub fn read_bytes<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    let current_bytes = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+        .unwrap()
+        .bytes()
+        .clone();
+    let position = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+        .unwrap()
+        .position();
+    let mut merging_offset = 0;
+    if let Some(Value::Object(second_array)) = args.get(0) {
+        let offset = args
+            .get(1)
+            .unwrap_or(&Value::Unsigned(0))
+            .coerce_to_u32(activation)? as usize;
+        let length = args
+            .get(2)
+            .unwrap_or(&Value::Unsigned(0))
+            .coerce_to_u32(activation)? as usize;
+
+        if position + length > current_bytes.len() {
+            log::error!("ByteArray: Reached EOF");
+            return Ok(Value::Undefined);
+        }
+        let mut merging_storage = second_array
+            .as_bytearray_mut(activation.context.gc_context)
+            .unwrap();
+
+        // Offset should not be greater then the buffer
+        if merging_storage.bytes().len() < offset {
+            return Ok(Value::Undefined);
+        }
+
+        let to_write = if length != 0 {
+            current_bytes[position..length + position].to_vec()
+        } else {
+            current_bytes[position..].to_vec()
+        };
+        merging_offset = to_write.len();
+        merging_storage.write_bytes_at(&to_write, offset);
+    }
+    this.unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+        .unwrap()
+        .add_position(merging_offset);
+    Ok(Value::Undefined)
+}
+pub fn write_utf<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Some(utf_string) = args.get(0) {
+            let utf_string = utf_string.coerce_to_string(activation)?;
+            bytearray.write_utf(&utf_string.as_str());
+        }
+    }
+    Ok(Value::Undefined)
+}
+
+pub fn read_utf<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
```

Comment:
You can avoid the unwrap and panic with `and_then`, like so: `this.and_then(|this| this.as_bytearray_mut(activation.context.gc_context))`

Reply:
I don't think that would work because the closure is returning data that is referencing `this` which is also owned by the function. Also I am a bit confused why this would fail anyways, if you could give an example that causes the issue, that would be very helpful